### DISCUSSION TITLE
使用html_sql的时候，针对sql不能换行

### DISCUSSION THREAD

**Discussion by AI1186780944** [rbatis/rbatis#308] [General]
使用的版本是rbatis 4.0.39。数据库是mysql。在书写长sql时，where之前的查询字段语句，需要进行回车sql换行，但是语句换行时会显示“/r”的回车符，从而导致sql错误。如果sql都不能换行，这一方面需要进行优化。

**Comment by zhuxiujia** [comment]
> 使用的版本是rbatis 4.0.39。数据库是mysql。在书写长sql时，where之前的查询字段语句，需要进行回车sql换行，但是语句换行时会显示“/r”的回车符，从而导致sql错误。如果sql都不能换行，这一方面需要进行优化。

支持换行在htmlsql的关键字 `` 里，例如
```rust
// Clion Smart tips: click code, choose 'Inject Language or Reference', and then choose html
#[html_sql(r#"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "https://raw.githubusercontent.com/rbatis/rbatis/master/rbatis-codegen/mybatis-3-mapper.dtd">
  <select id="select_by_condition">
        `select * from
        biz_activity`
  </select>"#)]
async fn select_by_condition(
    rb: &mut dyn Executor,
    name: &str,
    dt: &FastDateTime,
    a: bool,
) -> rbatis::Result<Vec<BizActivity>> {
    impled!()
}
```

```log
2022-10-23 19:05:40.9541424 INFO rbatis::plugin::log - [rbatis] [426698042314461184] Fetch  ==> select * from
        biz_activity
                                                      [rbatis]                      Args   ==> []
2022-10-23 19:05:40.9544611 INFO rbatis::plugin::log - [rbatis] [426698042314461184] ReturnRows <== 0
[]

```

**Reply by AI1186780944** [reply]
这是我的mapper：
```
#[html_sql("src/domain/html_sql/account.html")]
pub async fn select_by_user_no(rb: &mut dyn Executor,user_no:&str) -> Option<Account>{ impled!()}
```
这是我的sql语句：
```
<select id="select_by_user_no">
        `select
            id,usage_status,is_delete,create_id,create_time,
            update_id,update_time,parent_user_no,user_no,user_name,
            password,google_secret_key,nick_name,telephone,email,sex,company_no
        from account`
        <where>
            user_no = #{user_no}
        </where>
    </select>
```
这是报错信息：
```
2022-10-24 10:08:12.4854911 INFO rbatis::plugin::log - [rbatis] [426925170297540608] Fetch  ==> select\r
            id,usage_status,is_delete,create_id,create_time,\r
            update_id,update_time,parent_user_no,user_no,user_name,\r
            password,google_secret_key,nick_name,telephone,email,sex,company_no\r
        from account where user_no = ?
                                                      [rbatis]                      Args   ==> ["HCA1554083330340466688"]
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: E("1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '\\r\n            id,usage_status,is_delete,create_id,create_time,\\r\n            up' at line 1")', H:\rust program\recharge\src\infrastructure\repository\account_repository.rs:26:88
stack backtrace:
2022-10-24 10:08:12.49591 ERROR rbatis::plugin::log - [rbatis] [426925170297540608] ReturnError  <== 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '\r
            id,usage_status,is_delete,create_id,create_time,\r
            up' at line 1  C:\Users\11867\.cargo\registry\src\github.com-1ecc6299db9ec823\rbatis-4.0.39\src\plugin\log.rs:40
```
情形描述：当我按照你上面的方案，让sql语句换行并置于``的符号中，并且删除了target文件夹，重新进行了编译。执行main函数就会报这个sql语句，因为回车符导致的错误。

以下是不报错的书写案例：
```
<select id="select_by_user_no">
        `select id,usage_status,is_delete,create_id,create_time,update_id,update_time,parent_user_no,user_no,user_name,password,google_secret_key,nick_name,telephone,email,sex,company_no from account`
        <where>
            user_no = #{user_no}
        </where>
    </select>
```
控制台输出日志：
```
2022-10-24 10:21:47.4843221 INFO rbatis::plugin::log - [rbatis] [426928588651106304] Fetch  ==> select id,usage_status,is_delete,create_id,create_time,update_id,update_time,parent_user_no,user_no,user_name,password,google_secret_key,nick_name,telephone,email,sex,company_no from account where user_no = ?
                                                      [rbatis]                      Args   ==> ["HCA1554083330340466688"]
Some(AccountVo { id: Some("345c769bf0c2f336de294a1ef389a947"), usage_status: Some(1), is_delete: Some(0), create_id: Some("HCA1554083330340466688"), create_time: Some("2022-08-01 20:35:14.603"), update_id: Some("HCA1559149706624532480"), update_time: Some("2022-08-18 15:52:13.513"), parent_user_no: None, user_no: Some("HCA1554083330340466688"), user_name: Some("ai_seven"), password: Some("25d55ad283aa400af464c76d713c07ad"), google_secret_key: Some("M/lOGgV/1ccPhNvJHPYrGqIqSIv0FU8LiU1hkIXZAD0="), nick_name: Some("小七"), telephone: Some("18501763350"), email: Some("1186780944@qq.com"), sex: Some(1), company_no: Some("HCCC1554083330424352768") })
2022-10-24 10:21:47.5057353 INFO rbatis::plugin::log - [rbatis] [426928588651106304] ReturnRows <== 1
```

**Reply by zhuxiujia** [reply]
> 这是我的mapper： #[html_sql("src/domain/html_sql/account.html")] pub async fn select_by_user_no(rb: &mut dyn Executor,user_no:&str) -> Option{ impled!()} 这是我的sql语句： select id,usage_status,is_delete,create_id,create_time, update_id,update_time,parent_user_no,user_no,user_name, password,google_secret_key,nick_name,telephone,email,sex,company_no from account user_no = #{user_no} 这是报错信息： 2022-10-24 10:08:12.4854911 INFO rbatis::plugin::log - [rbatis] [426925170297540608] Fetch ==> select\r id,usage_status,is_delete,create_id,create_time,\r update_id,update_time,parent_user_no,user_no,user_name,\r password,google_secret_key,nick_name,telephone,email,sex,company_no\r from account where user_no = ? [rbatis] Args ==> ["HCA1554083330340466688"] thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: E("1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '\r\n id,usage_status,is_delete,create_id,create_time,\r\n up' at line 1")', H:\rust program\recharge\src\infrastructure\repository\account_repository.rs:26:88 stack backtrace: 2022-10-24 10:08:12.49591 ERROR rbatis::plugin::log - [rbatis] [426925170297540608] ReturnError <== 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '\r id,usage_status,is_delete,create_id,create_time,\r up' at line 1 C:\Users\11867.cargo\registry\src\github.com-1ecc6299db9ec823\rbatis-4.0.39\src\plugin\log.rs:40 情形描述：当我按照你上面的方案，让sql语句换行并置于``的符号中，并且删除了target文件夹，重新进行了编译。执行main函数就会报这个sql语句，因为回车符导致的错误。
> 
> 以下是不报错的书写案例： ![image](https://user-images.githubusercontent.com/40418583/197436278-1331cb0c-e326-4b91-8297-89ace04a5c2c.png) 控制台输出日志： 2022-10-24 10:21:47.4843221 INFO rbatis::plugin::log - [rbatis] [426928588651106304] Fetch ==> select id,usage_status,is_delete,create_id,create_time,update_id,update_time,parent_user_no,user_no,user_name,password,google_secret_key,nick_name,telephone,email,sex,company_no from account where user_no = ? [rbatis] Args ==> ["HCA1554083330340466688"] Some(AccountVo { id: Some("345c769bf0c2f336de294a1ef389a947"), usage_status: Some(1), is_delete: Some(0), create_id: Some("HCA1554083330340466688"), create_time: Some("2022-08-01 20:35:14.603"), update_id: Some("HCA1559149706624532480"), update_time: Some("2022-08-18 15:52:13.513"), parent_user_no: None, user_no: Some("HCA1554083330340466688"), user_name: Some("ai_seven"), password: Some("25d55ad283aa400af464c76d713c07ad"), google_secret_key: Some("M/lOGgV/1ccPhNvJHPYrGqIqSIv0FU8LiU1hkIXZAD0="), nick_name: Some("小七"), telephone: Some("18501763350"), email: Some("[1186780944@qq.com](mailto:1186780944@qq.com)"), sex: Some(1), company_no: Some("HCCC1554083330424352768") }) 2022-10-24 10:21:47.5057353 INFO rbatis::plugin::log - [rbatis] [426928588651106304] ReturnRows <== 1

不能使用/r吧，应该在``中直接输入回车键换行

**Reply by AI1186780944** [reply]
> > 这是我的mapper： #[html_sql("src/domain/html_sql/account.html")] pub async fn select_by_user_no(rb: &mut dyn Executor,user_no:&str) -> Option{ impled!()} 这是我的sql语句： select id,usage_status,is_delete,create_id,create_time, update_id,update_time,parent_user_no,user_no,user_name, password,google_secret_key,nick_name,telephone,email,sex,company_no from account user_no = #{user_no} 这是报错信息： 2022-10-24 10:08:12.4854911 INFO rbatis::plugin::log - [rbatis] [426925170297540608] Fetch ==> select\r id,usage_status,is_delete,create_id,create_time,\r update_id,update_time,parent_user_no,user_no,user_name,\r password,google_secret_key,nick_name,telephone,email,sex,company_no\r from account where user_no = ? [rbatis] Args ==> ["HCA1554083330340466688"] thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: E("1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '\r\n id,usage_status,is_delete,create_id,create_time,\r\n up' at line 1")', H:\rust program\recharge\src\infrastructure\repository\account_repository.rs:26:88 stack backtrace: 2022-10-24 10:08:12.49591 ERROR rbatis::plugin::log - [rbatis] [426925170297540608] ReturnError <== 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '\r id,usage_status,is_delete,create_id,create_time,\r up' at line 1 C:\Users\11867.cargo\registry\src\github.com-1ecc6299db9ec823\rbatis-4.0.39\src\plugin\log.rs:40 情形描述：当我按照你上面的方案，让sql语句换行并置于``的符号中，并且删除了target文件夹，重新进行了编译。执行main函数就会报这个sql语句，因为回车符导致的错误。
> > 以下是不报错的书写案例： ![image](https://user-images.githubusercontent.com/40418583/197436278-1331cb0c-e326-4b91-8297-89ace04a5c2c.png) 控制台输出日志： 2022-10-24 10:21:47.4843221 INFO rbatis::plugin::log - [rbatis] [426928588651106304] Fetch ==> select id,usage_status,is_delete,create_id,create_time,update_id,update_time,parent_user_no,user_no,user_name,password,google_secret_key,nick_name,telephone,email,sex,company_no from account where user_no = ? [rbatis] Args ==> ["HCA1554083330340466688"] Some(AccountVo { id: Some("345c769bf0c2f336de294a1ef389a947"), usage_status: Some(1), is_delete: Some(0), create_id: Some("HCA1554083330340466688"), create_time: Some("2022-08-01 20:35:14.603"), update_id: Some("HCA1559149706624532480"), update_time: Some("2022-08-18 15:52:13.513"), parent_user_no: None, user_no: Some("HCA1554083330340466688"), user_name: Some("ai_seven"), password: Some("25d55ad283aa400af464c76d713c07ad"), google_secret_key: Some("M/lOGgV/1ccPhNvJHPYrGqIqSIv0FU8LiU1hkIXZAD0="), nick_name: Some("小七"), telephone: Some("18501763350"), email: Some("[1186780944@qq.com](mailto:1186780944@qq.com)"), sex: Some(1), company_no: Some("HCCC1554083330424352768") }) 2022-10-24 10:21:47.5057353 INFO rbatis::plugin::log - [rbatis] [426928588651106304] ReturnRows <== 1
> 
> 不能使用/r吧，应该在``中直接输入回车键换行

问题就在这，我就是在编辑器（Clion）里用的回车键。然后最后识别出来的文本就自己带/r了。所以你看看是不是需要添加一个/r替换成/n的机制

**Reply by zhuxiujia** [reply]
我也是Clion，为啥我没有出现\r

**Reply by zhuxiujia** [reply]
我在Win和Mac的Clion敲入回车换行，都是跑通并且出现换行符的

**Reply by AI1186780944** [reply]
> 我在Win和Mac的Clion敲入回车换行，都是跑通并且出现换行符的

![image](https://user-images.githubusercontent.com/40418583/197437845-7c2f8b80-14ed-4518-9058-dd5de4b8fa08.png)
你自己看看

**Reply by AI1186780944** [reply]
> 我在Win和Mac的Clion敲入回车换行，都是跑通并且出现换行符的

![image](https://user-images.githubusercontent.com/40418583/197439391-aa43a616-44df-49f1-ba0c-34d8bfee0075.png)
我要是不换行，这条语句一点问题都没用。

**Reply by zhuxiujia** [reply]
你换vscode敲换行执行试试

**Reply by zhuxiujia** [reply]
![image](https://user-images.githubusercontent.com/3481148/197440979-3f3eff44-aa34-45dd-b0b3-c52ace342203.png)

**Reply by zhuxiujia** [reply]
我这边Clion执行正常

**Reply by AI1186780944** [reply]
> 你换vscode敲换行执行试试

![image](https://user-images.githubusercontent.com/40418583/197444847-dad3ab6e-2188-4a2a-a2bc-4fa36099456f.png)
换成vscode以后还是不行。请问当前环境使用的编码格式是什么？我的使用的是UTF-8

**Reply by AI1186780944** [reply]
> 我这边Clion执行正常
我重新构建一个新的项目来尝试运行带换行的sql也不行。还是会出现同样的问题。
